x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,212 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common/acker"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+)
+
+var ackerIDCounter *atomic.Uint64
+
+func init() {
+	ackerIDCounter = atomic.NewUint64(0)
+}
+
+// EventACKTracker tracks the publishing state of S3 objects. Specifically
+// it tracks the number of message acknowledgements that are pending from the
+// output. It can be used to wait until all ACKs have been received for one or
+// more S3 objects.
+type EventACKTracker struct {
+	ID uint64
+
+	EventsAcked       *atomic.Uint64
+	EventsDropped     *atomic.Uint64
+	EventsPublished   *atomic.Uint64
+	EventsToBeTracked *atomic.Uint64
+
+	ctx        context.Context
+	cancel     context.CancelFunc
+	deletionWg *sync.WaitGroup
+
+	mutex *sync.Mutex
+
+	msg             *types.Message
+	ReceiveCount    int
+	start           time.Time
+	processingErr   error
+	Handles         []s3ObjectHandler
+	keepaliveCancel context.CancelFunc
+	keepaliveWg     *sync.WaitGroup
+	msgHandler      sqsProcessor
+	log             *logp.Logger
+}
+
+func NewEventACKTracker(ctx context.Context, deletionWg *sync.WaitGroup) *EventACKTracker {
+	ctx, cancel := context.WithCancel(ctx)
+	acker := &EventACKTracker{
+		ID:                ackerIDCounter.Inc(),
+		ctx:               ctx,
+		cancel:            cancel,
+		deletionWg:        deletionWg,
+		mutex:             new(sync.Mutex),
+		EventsAcked:       atomic.NewUint64(0),
+		EventsDropped:     atomic.NewUint64(0),
+		EventsPublished:   atomic.NewUint64(0),
+		EventsToBeTracked: atomic.NewUint64(0),
+	}
+
+	go func() {
```

Comment:
Creating a looping goroutine for each message isn't needed. It should be enough to check `FullyTracked`:
- In `(*EventACKTracker).ACK`
- In `(*EventACKTracker).MarkSQSProcessedWithData`

since these are the only two places where the acked count and/or `EventsToBeTracked` can be modified. (This could be done safely with just atomics given some careful ordering, but it's also fine to use a mutex for the initial implementation, that's still a lot better than an extra goroutine for every message.)

Reply:
> Creating a looping goroutine for each message isn't needed. It should be enough to check `FullyTracked`:
> 
> * In `(*EventACKTracker).ACK`
> * In `(*EventACKTracker).MarkSQSProcessedWithData`

that's what we wanted to remove because it required a mutex, and it will block the listener callback.

Reply:
> In `EventPrivateReporter` I will just call `acker.ACK()` for every event, and in `LastEventPrivateReporter`:
> 
> ```
> if acker.FullyTracked() {
>     acker.FlushForSQS()
> }
> ```

beware that also this we can end up with the following timeline:
the timeline could be the following:
. T1 client publishes event 1
. T2 client publishes event 2
. T3 private event listener acks event 1
. T4 private event listener acks event 2
. T5 last event listener check how many events has to be acked (0, because T6 has to come) 
. T6 sqs event processor mark sqs message as processed, informing the acker that there are only 2 event

No more last event listener will be invocked. We don't delete the message and the input will stay appended on shutdown because we have `deletiongWg.Wait()`

So I guess a goroutine is needed indeed

Reply:
The core problem being that the s3 objects are read and published as a stream where you don't know how many events to expect before the object can be deleted until you have read them all?

https://github.com/aspacca/beats/blob/8c3c39a69efe506763aaad50efc5e8280cc0f7c5/x-pack/filebeat/input/awss3/s3_objects.go#L398-L403

```go
		message, err := reader.Next()
		if len(message.Content) > 0 {
			event := p.createEvent(string(message.Content), offset)
			event.Fields.DeepUpdate(message.Fields)
			offset += int64(message.Bytes)
			p.publish(&event)
		}

		if errors.Is(err, io.EOF) {
			// No more lines
			break
		}
```